### DISCUSSION TITLE
(Mac) AlwaysOnTop over fullscreen apps

### DISCUSSION THREAD

**Discussion by kroksys** [wailsapp/wails#2384] [Q&A]
Is it possible to put an app AlwaysOnTop over fullscreen apps? 

I think it would be possible using `NSFloatingWindowLevel` [developer.apple.com/documentation](https://developer.apple.com/documentation/appkit/nswindowlevel). Any ideas how could I get that functionality? I could make a branch and experiment, but C is not my strong side.

I would prefer to just manually choose it within wails/options/mac App config.

**Comment by misitebao** [comment]
This might help you.
https://wails.io/docs/next/reference/options#alwaysontop

**Reply by kroksys** [reply]
I tried `AlwaysOnTop: true` it works until you start to use fullscreen apps. There its not visible on top anymore.
As I understand from Apple doc the app needs to be started with [NSFloatingWindowLevel](https://developer.apple.com/documentation/appkit/nsfloatingwindowlevel) to honor `AlwaysOnTop` even over fullscreen apps.

**Reply by leaanthony** [reply]
Do you have a reference for that?

**Reply by kroksys** [reply]
I spend few hours trying to make proof of concept in Swift, but failed to do so (I dislike swift and apple docs)! I did get it onTop, but failed to make it visible over full-screen apps.

Its possible using Electron:
```
    app.dock.hide()
    mainWindow.setAlwaysOnTop(true, 'floating')
    mainWindow.setVisibleOnAllWorkspaces(true, { visibleOnFullScreen: true })
```

`visibleOnFullScreen` is implemented here:     [native_window_mac.mm](https://github.com/electron/electron/blob/9611b641c905f1eaa5e87c75d399004f28177167/shell/browser/native_window_mac.mm#L1290)

**Reply by leaanthony** [reply]
One possible solution is to set the window level above that of the fullscreen window. Something like:
```
    NSInteger floatingWindowLevel = CGWindowLevelForKey(kCGFloatingWindowLevelKey);
    [self.window setLevel:floatingWindowLevel + 1];
```
Or potentially we could just set it to MAX_INT or something.

**Reply by misitebao** [reply]
@kroksys 
If it is only running on your local computer, you can adjust the display mode of other full-screen applications to be borderless, which can achieve the effect you want. (Of course this is just a temporary solution😂).
